test/index.spec.ts

```diff
@@ -411,29 +447,52 @@ describe('[ TOTP ]', () => {
         },
         verify,
       )
-      const result = (await strategy
-        .authenticate(request, sessionStorage, {
-          ...AUTH_OPTIONS,
-          throwOnError: true,
-          successRedirect: '/',
+      let session: Session | undefined
+      {
+        const formData = new FormData()
+        formData.append(FORM_FIELDS.EMAIL, DEFAULT_EMAIL)
+        const request = new Request(`${HOST_URL}`, {
+          method: 'POST',
+          body: formData,
         })
-        .catch((error) => error)) as Response
+        await strategy
+          .authenticate(request, sessionStorage, {
+            ...AUTH_OPTIONS,
+            successRedirect: '/',
+          })
+          .catch(async (reason) => {
+            if (reason instanceof Response) {
+              expect(reason.status).toBe(302)
+              session = await sessionStorage.getSession(
+                reason.headers.get('set-cookie') ?? '',
+              )
+            } else throw reason
+          })
+      }
+      {
+        const totp = generateTOTP(TOTP_GENERATION_DEFAULTS)
+        const formData = new FormData()
+        formData.append(FORM_FIELDS.TOTP, totp.otp)
 
-      expect(result).toEqual(new AuthorizationError(ERRORS.TOTP_NOT_FOUND))
+        const request = new Request(`${HOST_URL}`, {
+          method: 'POST',
+          headers: {
+            cookie: await sessionStorage.commitSession(session!),
```

Comment:
Is there any way we could avoid using "!" in TypeScript code, @mw10013? I have a rule against it, and my editor is complaining badly.

- - - -

Here's a few solutions provided by Chat GPT:

The TypeScript error you're encountering, `Forbidden non-null assertion.eslint@typescript-eslint/no-non-null-assertion`, is related to the usage of the non-null assertion operator `!`. This operator is used to tell TypeScript that you are certain a value is not `null` or `undefined`.

In your code, you have `session!`. This tells TypeScript to assume `session` is not null or undefined. However, the TypeScript linter rule `@typescript-eslint/no-non-null-assertion` is set up to prevent this because non-null assertions can be unsafe. They bypass TypeScript's strict null checks, potentially leading to runtime errors if `session` is actually `null` or `undefined`.

To resolve this, you have a few options:

1. **Check for null/undefined**: Before using `session`, check if it's not `null` or `undefined`.

    ```typescript
    if (session) {
        cookie: await sessionStorage.commitSession(session),
    }
    ```

2. **Use Optional Chaining**: If `sessionStorage.commitSession` can handle `undefined`, use optional chaining.

    ```typescript
    cookie: await sessionStorage.commitSession(session ?? undefined),
    ```

3. **Provide a Default Value**: If there's a sensible default value for `session`, use it.

    ```typescript
    cookie: await sessionStorage.commitSession(session ?? defaultSession),
    ```

The best approach depends on your specific use case and how `session` is used in the rest of your code. Generally, it's best to handle potential `null` or `undefined` values explicitly to avoid runtime errors.